aclk/agent_cloud_link.c

```diff
@@ -893,6 +946,402 @@ static void aclk_main_cleanup(void *ptr)
     static_thread->enabled = NETDATA_MAIN_THREAD_EXITED;
 }
 
+int send_https_request(char *method, char *host, char *port, char *url, BUFFER *b, char *payload)
+{
+    struct timeval timeout = { .tv_sec = 30, .tv_usec = 0 };
+
+    size_t payload_len = 0;
+    if (payload != NULL)
+        payload_len = strlen(payload);
+
+    buffer_flush(b);
+    buffer_sprintf(
+        b,
+        "%s %s HTTP/1.1\r\nHost: %s\r\nAccept: plain/text\r\nContent-length: %zu\r\nAccept-Language: en-us\r\n"
+        "User-Agent: Netdata/rocks\r\n\r\n",
+        method, url, host, payload_len);
+    if (payload != NULL)
+        buffer_strcat(b, payload);
+    debug(D_ACLK, "Sending HTTPS req (%zu bytes): '%s'", b->len, buffer_tostring(b));
+    int sock = connect_to_this_ip46(IPPROTO_TCP, SOCK_STREAM, host, 0, port, &timeout);
+
+    if (unlikely(sock == -1)) {
+        error("Handshake failed");
+        return 1;
+    }
+
+    SSL_CTX *ctx = security_initialize_openssl_client();
+    // Certificate chain: not updating the stores - do we need private CA roots?
+    // Calls to SSL_CTX_load_verify_locations would go here.
+    SSL *ssl = SSL_new(ctx);
+    SSL_set_fd(ssl, sock);
+    int err = SSL_connect(ssl);
+    if (err!=1) {
+        error("SSL_connect() failed with err=%d", err);
+        SSL_free(ssl);
+        close(sock);
+        SSL_CTX_free(ctx);
+        return 1;
+    }
+    err = SSL_write(ssl, b->buffer, b->len);
+    if (err <= 0)
+    {
+        error("SSL_write() failed with err=%d", err);
+        SSL_free(ssl);
+        close(sock);
+        SSL_CTX_free(ctx);
+        return 1;
+    }
+    buffer_flush(b);
+    int bytes_read = SSL_read(ssl, b->buffer, b->size);
+    debug(D_ACLK, "Received %d bytes in response", bytes_read);
+    b->len = bytes_read;
+    SSL_shutdown(ssl);
+    SSL_free(ssl);
+    close(sock);
+        SSL_CTX_free(ctx);
+    return 0;
+}
+
+struct dictionary_singleton {
+    char *key;
+    char *result;
+};
+
+int json_extract_singleton(JSON_ENTRY *e)
+{
+    struct dictionary_singleton *data = e->callback_data;
+
+    switch (e->type) {
+        case JSON_OBJECT:
+        case JSON_ARRAY:
+            break;
+        case JSON_STRING:
+            if (!strcmp(e->name, data->key)) {
+                data->result = strdupz(e->data.string);
+                break;
+            }
+            break;
+        case JSON_NUMBER:
+        case JSON_BOOLEAN:
+        case JSON_NULL:
+            break;
+    }
+    return 0;
+}
+
+// Base-64 decoder.
+// Note: This is non-validating, invalid input will be decoded without an error.
+//       Challenges are packed into json strings so we don't skip newlines.
+//       Size errors (i.e. invalid input size or insufficient output space) are caught.
+size_t base64_decode(unsigned char *input, size_t input_size, unsigned char *output, size_t output_size)
+{
+    static char lookup[256];
+    static int first_time=1;
```

Comment:
I guess it's only being called by one thread?

Reply:
The initialization is idempotent but this is an interesting comment about the common idiom that we use all over the code.